### PR TITLE
joplin-desktop: add aarch64-darwin support

### PR DESCRIPTION
**Pull request by JesusMtnez**
###### Description of changes

Add support for `aarch64-darwin` systems.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/joplin-desktop/default.nix [hunk 1]
```diff
@@ -8,17 +8,11 @@ let
   inherit (stdenv.hostPlatform) system;
   throwSystem = throw "Unsupported system: ${system}";
 
-  suffix = {
-    x86_64-linux = "AppImage";
-    x86_64-darwin = "dmg";
-  }.${system} or throwSystem;
+  suffix = if stdenv.isDarwin then "dmg" else "AppImage";
```
Comment (wegank) [inline pkgs/applications/misc/joplin-desktop/default.nix:11]: Please edit on top of the original one.

File: pkgs/applications/misc/joplin-desktop/default.nix [hunk 2]
```diff
@@ -8,17 +8,11 @@ let
...
 
   src = fetchurl {
     url = "https://github.com/laurent22/joplin/releases/download/v${version}/Joplin-${version}.${suffix}";
-    sha256 = {
-      x86_64-linux = "sha256-oo3li8w1uem9lyFqwnrZ7Fl1R7Hrd8W+PHcIiaL2/+U=";
-      x86_64-darwin = "sha256-xYNp6WW8uPBrfuUgE5LI+1PuQK+vTA11eOtkz2ogpk0=";
-    }.${system} or throwSystem;
+    sha256 = if stdenv.isDarwin then "sha256-xYNp6WW8uPBrfuUgE5LI+1PuQK+vTA11eOtkz2ogpk0=" else "sha256-oo3li8w1uem9lyFqwnrZ7Fl1R7Hrd8W+PHcIiaL2/+U=";
```
Comment (wegank) [inline pkgs/applications/misc/joplin-desktop/default.nix:15]: Same here.

File: pkgs/applications/misc/joplin-desktop/default.nix (hunk 1 above)
Comment (JesusMtnez) [reply pkgs/applications/misc/joplin-desktop/default.nix:11]: @wegank `darwin` artifact is going to be the same for both architectures. Is it really necessary to duplicate it? I am open to better ways of doing this, but duplicating seems not the best one. What do you think?

Comment (wegank) [reply pkgs/applications/misc/joplin-desktop/default.nix:11]: You can do something like that with `suffix = rec {`:

https://github.com/NixOS/nixpkgs/blob/6e9531b9131da0c6542119f8d700e76d9c2eb1f4/pkgs/tools/networking/ookla-speedtest/default.nix#L28

Comment (JesusMtnez) [reply pkgs/applications/misc/joplin-desktop/default.nix]: I have re-write the changes to generate on `src` using the same strategy as in the example. It looks much cleaner, I think 🤔

**Comment by JesusMtnez** [comment]
@wegank @HugoReeves is there anything else missing to move this? Thank you

**Review by wegank — DISMISSED** [review]

**Comment by wegank** [comment]
Oh wait, it doesn't work on my aarch64-darwin machine...

**Comment by wegank** [comment]
Nope, there is no aarch64-darwin support...

```
$ file result/Applications/Joplin.app/Contents/MacOS/Joplin
result/Applications/Joplin.app/Contents/MacOS/Joplin: Mach-O 64-bit executable x86_64
```

**Comment by JesusMtnez** [comment]
I see, it seems aarch64-darwin support is still an ongoing issue https://github.com/laurent22/joplin/issues/6052, closing this for now.

**Pull request closed without merging**